core/src/proxy/auth.rs

```diff
@@ -42,206 +40,162 @@ impl Interceptor for AuthInterceptor {
+pub async fn generate_auth_tokens(
+    auth_service_client: &mut AuthServiceClient<Channel>,
+    // used to sign challenges
+    keypair: &Keypair,
+) -> crate::proxy::Result<(
+    Token, /* access_token */
+    Token, /* refresh_token */
+)> {
+    debug!("generate_auth_challenge");
+    let challenge_response = auth_service_client
+        .generate_auth_challenge(GenerateAuthChallengeRequest {
+            role: Role::Validator as i32,
+            pubkey: keypair.pubkey().as_ref().to_vec(),
+        })
+        .await
+        .map_err(|e| ProxyError::AuthenticationError(e.to_string()))?;
+
+    let formatted_challenge = format!(
+        "{}-{}",
+        keypair.pubkey(),
+        challenge_response.into_inner().challenge
+    );
+
+    let signed_challenge = keypair
+        .sign_message(formatted_challenge.as_bytes())
+        .as_ref()
+        .to_vec();
+
+    debug!(
+        "formatted_challenge: {} signed_challenge: {:?}",
+        formatted_challenge, signed_challenge
+    );
+
+    debug!("generate_auth_tokens");
+    let auth_tokens = auth_service_client
+        .generate_auth_tokens(GenerateAuthTokensRequest {
+            challenge: formatted_challenge,
+            client_pubkey: keypair.pubkey().as_ref().to_vec(),
+            signed_challenge,
+        })
+        .await
+        .map_err(|e| ProxyError::AuthenticationError(e.to_string()))?;
+
+    let inner = auth_tokens.into_inner();
+    let access_token = get_validated_token(inner.access_token)?;
+    let refresh_token = get_validated_token(inner.refresh_token)?;
+
+    Ok((access_token, refresh_token))
+}
 
-    /// Invokes the refresh_access_token gRPC method.
-    /// Returns a new access_token.
-    async fn refresh_access_token(
-        auth_service_client: &mut AuthServiceClient<Channel>,
-        refresh_token: Token,
-    ) -> crate::proxy::Result<Token> {
-        match auth_service_client
-            .refresh_access_token(RefreshAccessTokenRequest {
-                refresh_token: refresh_token.value,
-            })
-            .await
-        {
-            Ok(resp) => get_validated_token(resp.into_inner().access_token),
-            Err(e) => Err(ProxyError::GrpcError(e)),
-        }
-    }
+/// Tries to refresh the access token or run full-reauth if needed.
+/// This method writes to access_token if refresh is done.
+/// It overwrites the refresh token if full-reauth is run.
+pub async fn maybe_refresh_auth_tokens(
+    auth_service_client: &mut AuthServiceClient<Channel>,
+    datapoint_tokens_generated: &'static str,
+    datapoint_access_refresh: &'static str,
+    auth_uri: &String,
+    access_token: &Arc<Mutex<Token>>,
+    refresh_token: &mut Token,
+    cluster_info: &Arc<ClusterInfo>,
+    connection_timeout: &Duration,
+    auth_refresh_lookahead: u64,
+    num_full_refreshes: &mut u64,
+    num_refresh_access_token: &mut u64,
+) -> crate::proxy::Result<()> {
```

Comment:
is it possible to return (Option<Token>, Option<Token>) instead of passing in mutable access_token + refresh_token and metrics? pull metrics tracking to caller depending on what's returned

Reply:
mutating a bunch of state through parameters feels kinda icky